File: ai-microservice/app/ai_service.py
from .pinecone_service import PineconeService

class AiService:
    def __init__(self):
        self.pinecone_service = PineconeService()

    def ai_chat(self, user_query: str):
        vector = self.pinecone_service.get_embedding(user_query)
        search_results = self.pinecone_service.query_index(vector)

        results = []
        for match in search_results.get("matches", []):
            metadata = match.get("metadata", {})
            results.append({
                "id": match.get("id"),
                "name": metadata.get("name"),
                "price": metadata.get("price"),
                "description": metadata.get("description"),
                "thumbnailUrl": metadata.get("thumbnailUrl"),
                "createdAt": metadata.get("createdAt"),
                "score": match.get("score"),
            })

        message = (
            "It seems like the item you were looking for is available."
            if results else
            "It seems like the item you were looking for is unavailable."
        )

        return {
            "message": message,
            "items": results
        }

File: backend/src/core/ai/services/ai.service.ts
import { Injectable } from '@nestjs/common';
import { CreateAiDto } from '../dto/create-ai.dto';
import { UpdateAiDto } from '../dto/update-ai.dto';
import { Pinecone } from '@pinecone-database/pinecone';
import { OpenAIEmbeddings } from '@langchain/openai';

@Injectable()
export class AiService {
  public pinecone: Pinecone;
  public embedding: OpenAIEmbeddings;

  constructor() {
    this.pinecone = new Pinecone({ apiKey: process.env.PINECONE_API_KEY });
    this.embedding = new OpenAIEmbeddings({
      modelName: "text-embedding-ada-002",
      openAIApiKey: process.env.OPENAI_API_KEY,
    });
  }
  create(createAiDto: CreateAiDto) {
    return 'This action adds a new ai';
  }

  async aiChat(userQuery: string): Promise<any> {
    const index = this.pinecone.Index("exclusive-index");
    const userVector = await this.embedding.embedQuery(userQuery);

    const searchResults = await index.query({
        vector: userVector,
        topK: 3,
        includeMetadata: true,
    });

    const results = searchResults.matches.map(match => ({
        id: match.id,
        name: match.metadata.name,
        price: match.metadata.price,
        description: match.metadata.description,
        thumbnailUrl: match.metadata.thumbnailUrl,
        createdAt: match.metadata.createdAt,
        score: match.score,
    }));

    const message = results.length > 0 
        ? "It seems like the item you were looking for is available." 
        : "It seems like the item you were looking for is unavailable.";

    return {
        message,
        items: results,
    };
}

  findOne(id: number) {
    return `This action returns a #${id} ai`;
  }

  update(id: number, updateAiDto: UpdateAiDto) {
    return `This action updates a #${id} ai`;
  }

  remove(id: number) {
    return `This action removes a #${id} ai`;
  }
}
